### PR TITLE
feat(validator): parameter validation

### PR DESCRIPTION
**Pull request by soyuka**
Adds constraint to parameters to allow validation:

- [x] add tests
- [x] add a Metadata factory that automatically adds constraints based on OpenAPI and metadata informations

### CODE REVIEW

File: src/Metadata/Parameter.php [hunk 1]
```diff
@@ -89,6 +92,14 @@ public function getPriority(): ?int
         return $this->priority;
     }
 
+    /**
+     * @return Constraint|Constraint[]|null
+     */
+    public function getConstraint(): array|Constraint|null
```
Comment (vincentchalamon) [inline src/Metadata/Parameter.php:98]: I suggest to always manipulate an array of constraints

File: src/Symfony/Validator/State/ParameterValidatorProvider.php [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+use Symfony\Component\Validator\ConstraintViolation;
+use Symfony\Component\Validator\ConstraintViolationList;
+use Symfony\Component\Validator\Validator\ValidatorInterface;
+
+/**
+ * Validates parameters using the symfony validator.
+ *
+ * @experimental
+ */
+final class ParameterValidatorProvider implements ProviderInterface
```
Comment (vincentchalamon) [inline src/Symfony/Validator/State/ParameterValidatorProvider.php:28]:
```suggestion
final readonly class ParameterValidatorProvider implements ProviderInterface
```

File: src/Symfony/Validator/State/ParameterValidatorProvider.php [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+{
+    public function __construct(
+        private readonly ProviderInterface $decorated,
+        private readonly ValidatorInterface $validator
```
Comment (vincentchalamon) [inline src/Symfony/Validator/State/ParameterValidatorProvider.php:32]:
```suggestion
        private ProviderInterface $decorated,
        private ValidatorInterface $validator
```

File: src/Symfony/Validator/State/ParameterValidatorProvider.php [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    ) {
+    }
+
+    public function provide(Operation $operation, array $uriVariables = [], array $context = []): object|array|null
+    {
+        $body = $this->decorated->provide($operation, $uriVariables, $context);
+        if (!$context['request'] ?? null) {
```
Comment (vincentchalamon) [inline src/Symfony/Validator/State/ParameterValidatorProvider.php:39]:
```suggestion
        if (!$context['request'] instanceof Request) {
```

File: src/Symfony/Validator/State/ParameterValidatorProvider.php [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+            return $body;
+        }
+
+        $operation = $context['request']->attributes->get('_api_operation');
+        foreach ($operation->getParameters() as $parameter) {
+            if ($parameter->getRequired() && !\array_key_exists('_api_values', $parameter->getExtraProperties())) {
+                throw new ValidationException(new ConstraintViolationList([new ConstraintViolation(sprintf('Parameter "%s" is required.', $parameter->getKey()), null, [], null, null, null)]));
+            }
```
Comment (vincentchalamon) [inline src/Symfony/Validator/State/ParameterValidatorProvider.php:47]: Can't this constraint be automatically set in `ParameterResourceMetadataCollectionFactory` instead?

**Review by vincentchalamon — CHANGES_REQUESTED** [review]

File: src/Metadata/Resource/Factory/ParameterResourceMetadataCollectionFactory.php [hunk 6]
```diff
@@ -121,6 +142,69 @@ private function setDefaults(string $key, Parameter $parameter, string $resource
...
+
+        if ($openApi['uniqueItems'] ?? false) {
+            $assertions[] = new Unique();
+        }
+
+        if (isset($openApi['enum'])) {
+            $assertions[] = new Choice(choices: $openApi['enum']);
+        }
+
+        if (isset($openApi['enum'])) {
```
Comment (michaljusiega) [inline src/Metadata/Resource/Factory/ParameterResourceMetadataCollectionFactory.php:200]: Duplicated code.

File: src/Symfony/Validator/State/ParameterValidatorProvider.php (hunk 5 above)
Comment (soyuka) [reply src/Symfony/Validator/State/ParameterValidatorProvider.php:47]: Indeed, I'll try a default value to `null` with a NotBlank constraint?

File: src/Metadata/Parameter.php (hunk 1 above)
Comment (soyuka) [reply src/Metadata/Parameter.php:98]: For DX it's quite nice no? 

```
#[Parameter(constraint: new NotBlank())]
```

File: src/Symfony/Validator/State/ParameterValidatorProvider.php (hunk 5 above)
Comment (vincentchalamon) [reply src/Symfony/Validator/State/ParameterValidatorProvider.php:47]: Maybe just add NotBlank(allowNull: false) if `required` option is set?

File: src/Metadata/Parameter.php (hunk 1 above)
Comment (soyuka) [reply src/Metadata/Parameter.php:98]: it should rather be `new All(new NotBlank(), new Email())` so what about removing the `array`?

Comment (soyuka) [reply src/Metadata/Parameter.php:98]: actually `All` can't be used like this, we need to keep the array.

Comment (vincentchalamon) [reply src/Metadata/Parameter.php:98]: So `constraints`?
```
#[Parameter(constraints: [new NotBlank(), new Email()])
```

File: src/Symfony/Validator/State/ParameterValidatorProvider.php [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+use ApiPlatform\Metadata\Operation;
+use ApiPlatform\State\ProviderInterface;
+use ApiPlatform\Validator\Exception\ValidationException;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\Validator\ConstraintViolation;
+use Symfony\Component\Validator\ConstraintViolationList;
+use Symfony\Component\Validator\Validator\ValidatorInterface;
+
+/**
+ * Validates parameters using the symfony validator.
```
Comment (dunglas) [inline src/Symfony/Validator/State/ParameterValidatorProvider.php:25]:
```suggestion
 * Validates parameters using the Symfony validator.
```

File: src/Metadata/Parameter.php [hunk 8]
```diff
@@ -37,6 +39,7 @@ public function __construct(
         protected ?string $description = null,
         protected ?bool $required = null,
         protected ?int $priority = null,
+        protected Constraint|array|null $constraints = null,
```
Comment (dunglas) [inline src/Metadata/Parameter.php:42]: Shouldn't you cast the passed value to `array` in the constructor?

File: src/Metadata/Parameter.php [hunk 9]
```diff
@@ -178,6 +189,14 @@ public function withRequired(bool $required): static
         return $self;
     }
 
+    public function withConstraints(array|Constraint $constraints): static
+    {
+        $self = clone $this;
+        $self->constraints = $constraints;
```
Comment (dunglas) [inline src/Metadata/Parameter.php:195]:
```suggestion
        $self->constraints = (array) $constraints;
```

File: src/Metadata/Resource/Factory/ParameterResourceMetadataCollectionFactory.php [hunk 10]
```diff
@@ -121,6 +142,74 @@ private function setDefaults(string $key, Parameter $parameter, string $resource
...
 
+        $schema = $parameter->getSchema() ?? $parameter->getOpenApi()?->getSchema();
+        if (!$parameter->getConstraints()) {
+            $parameter = $this->addSchemaValidation($parameter, $schema, $parameter->getRequired() ?? $description['required'] ?? false, $parameter->getOpenApi());
+        }
+
         return $parameter;
     }
+
+    private function addSchemaValidation(Parameter $parameter, ?array $schema = null, bool $required = false, ?OpenApiParameter $openApi = null): Parameter
```
Comment (dunglas) [inline src/Metadata/Resource/Factory/ParameterResourceMetadataCollectionFactory.php:153]: Couldn't we make the dependency to `symfony/validator` optional, or even better, use inversion of control to move this method in the Symfony bridge (and add an extension point to allow the user to use another validation library)?

File: src/Metadata/Parameter.php [hunk 11]
```diff
@@ -89,6 +92,14 @@ public function getPriority(): ?int
         return $this->priority;
     }
 
+    /**
+     * @return Constraint|Constraint[]|null
```
[the remark this answers is not in the document]
Comment (soyuka) [reply src/Metadata/Parameter.php:96]: mhh why add the FQDN here?

File: src/Metadata/Parameter.php [hunk 12]
```diff
@@ -15,17 +15,19 @@
...
     /**
-     * @param array{type?: string}|null              $schema
-     * @param array<string, mixed>                   $extraProperties
-     * @param ProviderInterface|callable|string|null $provider
-     * @param FilterInterface|string|null            $filter
+     * @param array{type?: string}|null                                 $schema
+     * @param array<string, mixed>                                      $extraProperties
+     * @param ProviderInterface|callable|string|null                    $provider
+     * @param FilterInterface|string|null                               $filter
+     * @param Constraint|\Symfony\Component\Validator\Constraint[]|null $constraints
```
[the remark this answers is not in the document]
Comment (dunglas) [reply src/Metadata/Parameter.php:30]: I was mistaken by this. I thought we were also providing our own `Constraint` class but it's not the case.

So this should be: 

```suggestion
     * @param Constraint|Constraint[]|null $constraints
```

File: src/Metadata/Resource/Factory/ParameterResourceMetadataCollectionFactory.php (hunk 10 above)
Comment (dunglas) [reply src/Metadata/Resource/Factory/ParameterResourceMetadataCollectionFactory.php:153]: Ok, this is premature as we require the Symfony `Constraint` class anyway. Let's keep this as-is for now.

**Review by dunglas — APPROVED** [review]
LGTM when docs fix will be merged.

**Pull request merged**